### PR TITLE
tree-walk: optimize tree walk such that leaf detection of entries is delayed till the entry is sent on the treeWalkResult channel.

### PR DESCRIPTION
**Pull request by krishnasrinivas**
On minio2 for 100k entries (uuids as filenames)
`time mc ls myminio/testbucket/|head`
without delayIsLeaf optimization : 90 secs
with delayIsLeaf optimization: 5 secs

### CODE REVIEW

File: tree-walk.go [hunk 1]
```diff
@@ -37,43 +37,104 @@ type treeWalkResult struct {
...
 			if err == nil {
-				// Skip the entries which do not match the prefixEntry.
-				for i, entry := range entries {
-					if !strings.HasPrefix(entry, prefixEntry) {
-						entries[i] = ""
-						continue
+				// Sort and filter out the entries that does not have prefix prefixEntry.
+				sort.Strings(entries)
+				start := 0
+				end := len(entries)
```
Comment (krishnasrinivas) [inline tree-walk.go]: @harshavardhana haven't tried to add comments for the code below as code explains the logic better than words. have added the gist `Sort and filter out the entries that does not have prefix prefixEntry.` above

**Comment by harshavardhana** [comment]
Can you add test cases as well for various styles of input for 'delayIsLeafCheck()' ?

File: tree-walk.go [hunk 2]
```diff
@@ -37,43 +37,104 @@ type treeWalkResult struct {
...
+				for {
+					if start == end {
+						break
 					}
```
Comment (harshavardhana) [inline tree-walk.go]: This code block has become bigger, we should move this a function which basically just skips all entries which do not have the prefix.

Comment (harshavardhana) [inline tree-walk.go]: > This code block has become bigger, we should move this a function which basically just skips all entries which do not have the prefix.

Along with a test case as well.

**Comment by harshavardhana** [comment]
> On minio2 for 100k entries (uuids as filenames)
> time mc ls myminio/testbucket/|head
> without delayIsLeaf optimization : 90 secs
> with delayIsLeaf optimization: 5 secs

Can you do more testing ? basically provide a more average baseline here. Would be nice to know the overall benefit gradient with 
- 100k, 500k, 1mil, 2mil in same directory. 
- 100k, 500k, 1mil, 2mil spread out.

**Comment by krishnasrinivas** [comment]
> > On minio2 for 100k entries (uuids as filenames)
> > time mc ls myminio/testbucket/|head
> > without delayIsLeaf optimization : 90 secs
> > with delayIsLeaf optimization: 5 secs
> 
> Can you do more testing ? basically provide a more average baseline here. Would be nice to know the overall benefit gradient with
> 
> 100k, 500k, 1mil, 2mil in same directory.
> 100k, 500k, 1mil, 2mil spread out.

From previous experience, the numbers can be extrapolated, but I will check it once again.

File: tree-walk.go [hunk 3]
```diff
@@ -37,43 +37,108 @@ type treeWalkResult struct {
...
+// Hence if entries[] does not have a case like the above example then isLeaf() check
+// can be delayed till the entry is pushed into the treeWalkResult channel.
+// delayIsLeafCheck() returns true if isLeaf can be delayed or false if
+// isLeaf should be done in listDir()
+func delayIsLeafCheck(entries []string) bool {
+	for i, entry := range entries {
+		if i == len(entries)-1 {
+			break
+		}
+		// For each rune in the entry string, if any rune is less than '/'
```
Comment (krisis) [inline tree-walk.go]: `entry[j] < '/'` is performing a byte comparison and not a rune comparison. Please update the comment to reflect the same.

**Comment by harshavardhana** [comment]
This patch can be accepted.. just that while we are at it we could finish it.

File: tree-walk.go (hunk 3 above)
Comment (krishnasrinivas) [inline tree-walk.go]: ah, I thought indexing into a string fetches a rune and not byte, but you are right:

```
krishna@escape:/tmp$ cat a.go 
package main

import "fmt"

func main() {
    a := "⌘"
    for i := range a {
        fmt.Println(a[i])
    }
    fmt.Println("====")
    b := []byte(a)
    for i := range b {
        fmt.Println(b[i])
    }
}
krishna@escape:/tmp$ go run a.go
226
====
226
140
152
krishna@escape:/tmp$ 
```

**Comment by krishnasrinivas** [comment]
> This patch can be accepted.. just that while we are at it we could finish it.

In that case you can take the patch in, I am confident that the numbers can be extrapolated. For majority of the cases, the only overhead with the current patch is delayIsLeafCheck which was quick on my laptop (2 secs for 10 mil entries)

I will do  performance testing for listing later in the week once I get other things off my list.

**Comment by harshavardhana** [comment]
LGTM, Thanks :+1:

**Pull request merged**